src/plaquette/device/_stimsim.py

```diff
@@ -112,19 +116,8 @@ def __init__(
         reason, :attr:`get_sample()` pre-computes a number of ``batch_size`` samples
         whenever it needs to get new samples from Stim.
         """
-        if isinstance(circ, circuit.CircuitBuilder):
-            self.circ = circ.circ
-        elif isinstance(circ, circuit.Circuit):
-            self.circ = circ
-        else:
-            raise TypeError(
-                "Only a Circuit or a CircuitBuilder can be used in a simulator"
-            )
         self.batch_size = batch_size
-        # Convert the circuit to Stim format. A sampler is not built until an RNG
-        # is supplied.
-        self.stim_circ, is_erasure = circuit_to_stim(self.circ)
-        self.meas_is_erasure = np.array(is_erasure)
+
```

Comment:
```suggestion
        self.circ: plaq_circuit.Circuit | None = None
        self.stim_circ: stim.Circuit | None = None
        self.meas_is_erasure = np.ndarray | None = None
        self.stim_sampler: stim.CompiledMeasurementSampler | None = None

```

Always init all instance variables in the init.

Reply:
* My thought here was that since we're calling `self.reset`, we could break the conventional initialization of instance variables in `__init__`. If we do want to do that nonetheless, there's going to be a bit of a duplication between `__init__` and `reset`, are we okay with that?
* Setting `self.circ` to `None` ends up bringing up an issue with MyPy:
```
src/plaquette/device/_stimsim.py:120: error: Incompatible types in assignment (expression has type "None", variable has type "Circuit")  [assignment]
```
Frankly, I'm not sure where this is coming from because it persists even if I remove any uses of `self.circ`. A workaround was to do `self.circ = plaq_circuit.Circuit()` as the initialization step.

Reply:
> we do want to do that nonetheless, there's going to be a bit of a duplication between __init__ and reset, are we okay with that?

It's not a problem, because worse IMO is the additional mental overhead of having to realise that some variables appear in the class that you never saw in the init. It's not strictly repetition, because resetting is a legitimate operation: all variables are reset to a known state. Initialisation is different semantically, although practically they do basically the exact same things. Some more strict linters (e.g. I think Microsoft's Pyright) actually mark this as an error.

> Setting self.circ to None ends up bringing up an issue with MyPy

This is due to a wrong way of declaring the types of class variables. You *could* declare all your instance variables as class variables (lines 78-95), but this ends up being a mess and it should actually be removed. It was a spread convention in the codebase that was only bringing in troubles with type checkers and linters (as you also see), this in the `stim` simulator must have slipped through the cracks.

The solution is to declare instance variables *and their types* in a single spot: the `__init__`. For example, if you delete L84, mypy stops complaining about that (but starts complaining about other things). Usually fixing `mypy` complaints, even if it seems verbose, leads to less bugs, but of course it cannot become a dogma to follow. We should just use common sense, e.g. if fixing mypy complaints ends up adding 200 lines of code, maybe we can simply ignore some stuff, after careful checking.

Reply:
> It's not a problem, because worse IMO is the additional mental overhead of having to realise that some variables appear in the class that you never saw in the init. It's not strictly repetition, because resetting is a legitimate operation: all variables are reset to a known state. Initialisation is different semantically, although practically they do basically the exact same things. Some more strict linters (e.g. I think Microsoft's Pyright) actually mark this as an error.

Oh, I see! Frankly, I've looked at this from a different angle where `__init__` and `reset` together are pretty small (hopefully) so it wouldn't bother the reader if something is only initialized in `reset`. Having said that, the suggestion is definitely more conventional and will likely benefit readability, so I've applied the changes.

> This is due to a wrong way of declaring the types of class variables. You could declare all your instance variables as class variables (lines 78-95), but this ends up being a mess and it should actually be removed. It was a spread convention in the codebase that was only bringing in troubles with type checkers and linters (as you also see), this in the stim simulator must have slipped through the cracks.

> The solution is to declare instance variables and their types in a single spot: the __init__. For example, if you delete L84, mypy stops complaining about that (but starts complaining about other things). Usually fixing mypy complaints, even if it seems verbose, leads to less bugs, but of course it cannot become a dogma to follow. We should just use common sense, e.g. if fixing mypy complaints ends up adding 200 lines of code, maybe we can simply ignore some stuff, after careful checking.

Oh wow, thanks for pointing that out! I've removed the declarations of the instance attributes on the class level and moved them into `__init__`. Also, I've changed to the preferred documentation of the attributes - although out of scope here - to make things simpler.